// vite-plugin-inspect/dist/node-BulLjr0V.mjs
import { DIR_CLIENT } from "./dirs.mjs";
import c from "ansis";
import { debounce } from "perfect-debounce";
import sirv from "sirv";
import { isAbsolute, join, resolve } from "node:path";
import fs from "node:fs/promises";
import process from "node:process";
import { hash } from "ohash";
import { Buffer } from "node:buffer";
import { createFilter } from "unplugin-utils";
import { parse } from "error-stack-parser-es";
import { createDebug } from "obug";
import { defineRpcFunction } from "@vitejs/devtools-kit";
//#region src/node/build.ts
function createEnvOrderHooks(environmentNames, { onFirst, onEach, onLast }) {
	const remainingEnvs = new Set(environmentNames);
	let ranFirst = false;
	let ranLast = false;
	return async (envName, ...args) => {
		if (!ranFirst) {
			ranFirst = true;
			await (onFirst === null || onFirst === void 0 ? void 0 : onFirst(...args));
		}
		remainingEnvs.delete(envName);
		await (onEach === null || onEach === void 0 ? void 0 : onEach(...args));
		if (!ranLast && remainingEnvs.size === 0) {
			ranLast = true;
			await (onLast === null || onLast === void 0 ? void 0 : onLast(...args));
		}
	};
}
function createBuildGenerator(ctx) {
	const { outputDir = ".vite-inspect" } = ctx.options;
	const targetDir = isAbsolute(outputDir) ? outputDir : resolve(process.cwd(), outputDir);
	const reportsDir = join(targetDir, "reports");
	return {
		getOutputDir() {
			return targetDir;
		},
		async setupOutputDir() {
			await fs.rm(targetDir, {
				recursive: true,
				force: true
			});
			await fs.mkdir(reportsDir, { recursive: true });
			await fs.cp(DIR_CLIENT, targetDir, { recursive: true });
		},
		async generateForEnv(pluginCtx) {
			const env = pluginCtx.environment;
			await Promise.all([...ctx._idToInstances.values()].filter((v) => v.environments.has(env.name)).map((v) => {
				const e = v.environments.get(env.name);
				return [`${v.id}-${env.name}`, e];
			}).map(async ([key, env]) => {
				await fs.mkdir(join(reportsDir, key, "transforms"), { recursive: true });
				return await Promise.all([
					writeJSON(join(reportsDir, key, "modules.json"), env.getModulesList(pluginCtx)),
					writeJSON(join(reportsDir, key, "metric-plugins.json"), env.getPluginMetrics()),
					...Object.entries(env.data.transform).map(([id, info]) => writeJSON(join(reportsDir, key, "transforms", `${hash(id)}.json`), {
						resolvedId: id,
						transforms: info
					}))
				]);
			}));
		},
		async generateMetadata() {
			await writeJSON(join(reportsDir, "metadata.json"), ctx.getMetadata());
		}
	};
}
function writeJSON(filename, data) {
	return fs.writeFile(filename, `${JSON.stringify(data, null, 2)}\n`);
}
//#endregion
//#region src/node/constants.ts
const DUMMY_LOAD_PLUGIN_NAME = "__load__";
//#endregion
//#region src/node/utils.ts
function serializePlugin(plugin) {
	return JSON.parse(JSON.stringify(plugin, (key, value) => {
		if (typeof value === "function") {
			let name = value.name;
			if (name === "anonymous") name = "";
			if (name === key) name = "";
			if (name) return `[Function ${name}]`;
			return "[Function]";
		}
		if (key === "api" && value) return "[Object API]";
		return value;
	}));
}
function removeVersionQuery(url) {
	if (url.includes("v=")) return url.replace(/&v=\w+/, "").replace(/\?v=\w+/, "?").replace(/\?$/, "");
	return url;
}
//#endregion
//#region src/node/context.ts
let viteCount = 0;
var InspectContext = class {
	_configToInstances = /* @__PURE__ */ new Map();
	_idToInstances = /* @__PURE__ */ new Map();
	filter;
	constructor(options) {
		this.options = options;
		this.filter = createFilter(options.include, options.exclude);
	}
	getMetadata() {
		return {
			instances: Array.from(this._idToInstances.values(), (vite) => ({
				root: vite.config.root,
				vite: vite.id,
				plugins: vite.config.plugins.map((i) => serializePlugin(i)),
				environments: [...vite.environments.keys()],
				environmentPlugins: Object.fromEntries(Array.from(vite.environments.entries(), ([name, env]) => {
					return [name, env.env.getTopLevelConfig().plugins.map((i) => vite.config.plugins.indexOf(i))];
				}))
			})),
			embedded: this.options.embedded
		};
	}
	getViteContext(configOrId) {
		if (typeof configOrId === "string") {
			if (!this._idToInstances.has(configOrId)) throw new Error(`Can not found vite context for ${configOrId}`);
			return this._idToInstances.get(configOrId);
		}
		if (this._configToInstances.has(configOrId)) return this._configToInstances.get(configOrId);
		const id = `vite${++viteCount}`;
		const vite = new InspectContextVite(id, this, configOrId);
		this._idToInstances.set(id, vite);
		this._configToInstances.set(configOrId, vite);
		return vite;
	}
	getEnvContext(env) {
		if (!env) return void 0;
		return this.getViteContext(env.getTopLevelConfig()).getEnvContext(env);
	}
	queryEnv(query) {
		return this.getViteContext(query.vite).getEnvContext(query.env);
	}
};
var InspectContextVite = class {
	environments = /* @__PURE__ */ new Map();
	data = { serverMetrics: { middleware: {} } };
	constructor(id, context, config) {
		this.id = id;
		this.context = context;
		this.config = config;
	}
	getEnvContext(env) {
		if (typeof env === "string") {
			if (!this.environments.has(env)) throw new Error(`Can not found environment context for ${env}`);
			return this.environments.get(env);
		}
		if (env.getTopLevelConfig() !== this.config) throw new Error("Environment config does not match Vite config");
		if (!this.environments.has(env.name)) this.environments.set(env.name, new InspectContextViteEnv(this.context, this, env));
		return this.environments.get(env.name);
	}
};
var InspectContextViteEnv = class {
	constructor(contextMain, contextVite, env) {
		this.contextMain = contextMain;
		this.contextVite = contextVite;
		this.env = env;
	}
	data = {
		transform: {},
		resolveId: {},
		transformCounter: {}
	};
	recordTransform(id, info, preTransformCode) {
		id = this.normalizeId(id);
		if (!this.data.transform[id] || !this.data.transform[id].some((tr) => tr.result)) {
			this.data.transform[id] = [{
				name: DUMMY_LOAD_PLUGIN_NAME,
				result: preTransformCode,
				start: info.start,
				end: info.start,
				sourcemaps: info.sourcemaps
			}];
			this.data.transformCounter[id] = (this.data.transformCounter[id] || 0) + 1;
		}
		this.data.transform[id].push(info);
	}
	recordLoad(id, info) {
		id = this.normalizeId(id);
		this.data.transform[id] = [info];
		this.data.transformCounter[id] = (this.data.transformCounter[id] || 0) + 1;
	}
	recordResolveId(id, info) {
		id = this.normalizeId(id);
		if (!this.data.resolveId[id]) this.data.resolveId[id] = [];
		this.data.resolveId[id].push(info);
	}
	invalidate(id) {
		id = this.normalizeId(id);
		delete this.data.transform[id];
	}
	normalizeId(id) {
		if (this.contextMain.options.removeVersionQuery !== false) return removeVersionQuery(id);
		return id;
	}
	getModulesList(pluginCtx) {
		const moduleGraph = this.env.mode === "dev" ? this.env.moduleGraph : void 0;
		const getDeps = moduleGraph ? (id) => {
			var _moduleGraph$getModul;
			return Array.from(((_moduleGraph$getModul = moduleGraph.getModuleById(id)) === null || _moduleGraph$getModul === void 0 ? void 0 : _moduleGraph$getModul.importedModules) || []).map((i) => i.id || "").filter(Boolean);
		} : pluginCtx ? (id) => {
			var _pluginCtx$getModuleI;
			return ((_pluginCtx$getModuleI = pluginCtx.getModuleInfo(id)) === null || _pluginCtx$getModuleI === void 0 ? void 0 : _pluginCtx$getModuleI.importedIds) || [];
		} : () => [];
		const getImporters = moduleGraph ? (id) => {
			var _moduleGraph$getModul2;
			return Array.from((moduleGraph === null || moduleGraph === void 0 || (_moduleGraph$getModul2 = moduleGraph.getModuleById(id)) === null || _moduleGraph$getModul2 === void 0 ? void 0 : _moduleGraph$getModul2.importers) || []).map((i) => i.id || "").filter(Boolean);
		} : pluginCtx ? (id) => {
			var _pluginCtx$getModuleI2;
			return ((_pluginCtx$getModuleI2 = pluginCtx.getModuleInfo(id)) === null || _pluginCtx$getModuleI2 === void 0 ? void 0 : _pluginCtx$getModuleI2.importers) || [];
		} : () => [];
		function isVirtual(pluginName, transformName) {
			return pluginName !== "__load__" && transformName !== "vite:load-fallback" && transformName !== "vite:build-load-fallback";
		}
		const transformedIdMap = Object.values(this.data.resolveId).reduce((map, ids) => {
			ids.forEach((id) => {
				var _id$result;
				map[_id$result = id.result] ?? (map[_id$result] = []);
				map[id.result].push(id);
			});
			return map;
		}, {});
		const ids = new Set(Object.keys(this.data.transform).concat(Object.keys(transformedIdMap)));
		return Array.from(ids).sort().map((id) => {
			var _plugins$, _this$data$transform$, _this$data$transformC, _this$data$transform$2, _this$data$transform$3;
			let totalTime = 0;
			const plugins = (this.data.transform[id] || []).filter((tr) => tr.result).map((transItem) => {
				const delta = transItem.end - transItem.start;
				totalTime += delta;
				return {
					name: transItem.name,
					transform: delta
				};
			}).concat((transformedIdMap[id] || []).map((idItem) => {
				return {
					name: idItem.name,
					resolveId: idItem.end - idItem.start
				};
			}));
			function getSize(str) {
				if (!str) return 0;
				return Buffer.byteLength(str, "utf8");
			}
			return {
				id,
				deps: getDeps(id),
				importers: getImporters(id),
				plugins,
				virtual: isVirtual(((_plugins$ = plugins[0]) === null || _plugins$ === void 0 ? void 0 : _plugins$.name) || "", ((_this$data$transform$ = this.data.transform[id]) === null || _this$data$transform$ === void 0 ? void 0 : _this$data$transform$[0].name) || ""),
				totalTime,
				invokeCount: ((_this$data$transformC = this.data.transformCounter) === null || _this$data$transformC === void 0 ? void 0 : _this$data$transformC[id]) || 0,
				sourceSize: getSize((_this$data$transform$2 = this.data.transform[id]) === null || _this$data$transform$2 === void 0 || (_this$data$transform$2 = _this$data$transform$2[0]) === null || _this$data$transform$2 === void 0 ? void 0 : _this$data$transform$2.result),
				distSize: getSize((_this$data$transform$3 = this.data.transform[id]) === null || _this$data$transform$3 === void 0 || (_this$data$transform$3 = _this$data$transform$3.at(-1)) === null || _this$data$transform$3 === void 0 ? void 0 : _this$data$transform$3.result)
			};
		});
	}
	resolveId(id = "") {
		if (id.startsWith("./")) id = resolve(this.env.getTopLevelConfig().root, id).replace(/\\/g, "/");
		return this.resolveIdRecursive(id);
	}
	resolveIdRecursive(id) {
		var _this$data$resolveId$;
		const resolved = (_this$data$resolveId$ = this.data.resolveId[id]) === null || _this$data$resolveId$ === void 0 || (_this$data$resolveId$ = _this$data$resolveId$[0]) === null || _this$data$resolveId$ === void 0 ? void 0 : _this$data$resolveId$.result;
		return resolved ? this.resolveIdRecursive(resolved) : id;
	}
	getPluginMetrics() {
		const map = {};
		const defaultMetricInfo = () => ({
			transform: {
				invokeCount: 0,
				totalTime: 0
			},
			resolveId: {
				invokeCount: 0,
				totalTime: 0
			}
		});
		this.env.getTopLevelConfig().plugins.forEach((i) => {
			map[i.name] = {
				...defaultMetricInfo(),
				name: i.name,
				enforce: i.enforce
			};
		});
		Object.values(this.data.transform).forEach((transformInfos) => {
			transformInfos.forEach(({ name, start, end }) => {
				if (name === "__load__") return;
				if (!map[name]) map[name] = {
					...defaultMetricInfo(),
					name
				};
				map[name].transform.totalTime += end - start;
				map[name].transform.invokeCount += 1;
			});
		});
		Object.values(this.data.resolveId).forEach((resolveIdInfos) => {
			resolveIdInfos.forEach(({ name, start, end }) => {
				if (!map[name]) map[name] = {
					...defaultMetricInfo(),
					name
				};
				map[name].resolveId.totalTime += end - start;
				map[name].resolveId.invokeCount += 1;
			});
		});
		return Object.values(map).filter(Boolean).sort((a, b) => a.name.localeCompare(b.name));
	}
	async getModuleTransformInfo(id) {
		const resolvedId = this.resolveId(id);
		return {
			resolvedId,
			transforms: this.data.transform[resolvedId] || []
		};
	}
	async clearModuleTransform(id) {
		this.clearId(id);
		try {
			if (this.env.mode === "dev") await this.env.transformRequest(id);
		} catch {}
	}
	clearId(_id) {
		const id = this.resolveId(_id);
		if (id) {
			const moduleGraph = this.env.mode === "dev" ? this.env.moduleGraph : void 0;
			const mod = moduleGraph === null || moduleGraph === void 0 ? void 0 : moduleGraph.getModuleById(id);
			if (mod) moduleGraph === null || moduleGraph === void 0 || moduleGraph.invalidateModule(mod);
			this.invalidate(id);
		}
	}
};
//#endregion
//#region src/node/hijack.ts
const debug = createDebug("vite-plugin-inspect");
function hijackHook(plugin, name, wrapper) {
	if (!plugin[name]) return;
	debug(`hijack plugin "${name}"`, plugin.name);
	let order = plugin.order || plugin.enforce || "normal";
	const hook = plugin[name];
	if ("handler" in hook) {
		const oldFn = hook.handler;
		order += `-${hook.order || hook.enforce || "normal"}`;
		hook.handler = function(...args) {
			return wrapper(oldFn, this, args, order);
		};
	} else if ("transform" in hook) {
		const oldFn = hook.transform;
		order += `-${hook.order || hook.enforce || "normal"}`;
		hook.transform = function(...args) {
			return wrapper(oldFn, this, args, order);
		};
	} else {
		const oldFn = hook;
		plugin[name] = function(...args) {
			return wrapper(oldFn, this, args, order);
		};
	}
}
const hijackedPlugins = /* @__PURE__ */ new WeakSet();
function hijackPlugin(plugin, ctx) {
	if (hijackedPlugins.has(plugin)) return;
	hijackedPlugins.add(plugin);
	hijackHook(plugin, "transform", async (fn, context, args, order) => {
		const code = args[0];
		const id = args[1];
		let _result;
		let error;
		const start = Date.now();
		try {
			_result = await fn.apply(context, args);
		} catch (_err) {
			error = _err;
		}
		const end = Date.now();
		const result = error ? "[Error]" : typeof _result === "string" ? _result : _result === null || _result === void 0 ? void 0 : _result.code;
		if (ctx.filter(id)) {
			var _ctx$getEnvContext;
			const sourcemaps = typeof _result === "string" ? null : _result === null || _result === void 0 ? void 0 : _result.map;
			(_ctx$getEnvContext = ctx.getEnvContext(context === null || context === void 0 ? void 0 : context.environment)) === null || _ctx$getEnvContext === void 0 || _ctx$getEnvContext.recordTransform(id, {
				name: plugin.name,
				result,
				start,
				end,
				order,
				sourcemaps,
				error: error ? parseError(error) : void 0
			}, code);
		}
		if (error) throw error;
		return _result;
	});
	hijackHook(plugin, "load", async (fn, context, args) => {
		const id = args[0];
		let _result;
		let error;
		const start = Date.now();
		try {
			_result = await fn.apply(context, args);
		} catch (err) {
			error = err;
		}
		const end = Date.now();
		const result = error ? "[Error]" : typeof _result === "string" ? _result : _result === null || _result === void 0 ? void 0 : _result.code;
		const sourcemaps = typeof _result === "string" ? null : _result === null || _result === void 0 ? void 0 : _result.map;
		if (result) {
			var _ctx$getEnvContext2;
			(_ctx$getEnvContext2 = ctx.getEnvContext(context === null || context === void 0 ? void 0 : context.environment)) === null || _ctx$getEnvContext2 === void 0 || _ctx$getEnvContext2.recordLoad(id, {
				name: plugin.name,
				result,
				start,
				end,
				sourcemaps,
				error: error ? parseError(error) : void 0
			});
		}
		if (error) throw error;
		return _result;
	});
	hijackHook(plugin, "resolveId", async (fn, context, args) => {
		const id = args[0];
		let _result;
		let error;
		const start = Date.now();
		try {
			_result = await fn.apply(context, args);
		} catch (err) {
			error = err;
		}
		const end = Date.now();
		if (!ctx.filter(id)) {
			if (error) throw error;
			return _result;
		}
		const result = error ? stringifyError(error) : typeof _result === "object" ? _result === null || _result === void 0 ? void 0 : _result.id : _result;
		if (result && result !== id) {
			var _ctx$getEnvContext3;
			(_ctx$getEnvContext3 = ctx.getEnvContext(context === null || context === void 0 ? void 0 : context.environment)) === null || _ctx$getEnvContext3 === void 0 || _ctx$getEnvContext3.recordResolveId(id, {
				name: plugin.name,
				result,
				start,
				end,
				error
			});
		}
		if (error) throw error;
		return _result;
	});
}
function parseError(error) {
	const stack = parse(error, { allowEmpty: true });
	return {
		message: error.message || String(error),
		stack,
		raw: error
	};
}
function stringifyError(err) {
	return String(err.stack ? err.stack : err);
}
//#endregion
//#region src/node/rpc/ctx.ts
const contextMap = /* @__PURE__ */ new WeakMap();
function setInspectContext(devtoolsCtx, ctx) {
	contextMap.set(devtoolsCtx, ctx);
}
function getInspectContext(devtoolsCtx) {
	const ctx = contextMap.get(devtoolsCtx);
	if (!ctx) throw new Error("InspectContext not found for this DevTools context");
	return ctx;
}
//#endregion
//#region src/node/rpc/functions/clear-module-transform.ts
const clearModuleTransform = defineRpcFunction({
	name: "vite-plugin-inspect:clear-module-transform",
	type: "action",
	setup: (devtoolsCtx) => {
		const ctx = getInspectContext(devtoolsCtx);
		return { handler: async (query, id) => {
			await ctx.queryEnv(query).clearModuleTransform(id);
		} };
	}
});
//#endregion
//#region src/node/rpc/functions/get-metadata.ts
const getMetadata = defineRpcFunction({
	name: "vite-plugin-inspect:get-metadata",
	type: "query",
	setup: (devtoolsCtx) => {
		const ctx = getInspectContext(devtoolsCtx);
		return {
			handler: async () => ctx.getMetadata(),
			dump: { inputs: [[]] }
		};
	}
});
//#endregion
//#region src/node/rpc/utils.ts
function getAllQueryEnvs(ctx) {
	const result = [];
	for (const vite of ctx._idToInstances.values()) for (const envName of vite.environments.keys()) result.push({
		vite: vite.id,
		env: envName
	});
	return result;
}
function getAllModuleIds(ctx) {
	const result = [];
	for (const vite of ctx._idToInstances.values()) for (const [envName, env] of vite.environments) {
		const query = {
			vite: vite.id,
			env: envName
		};
		for (const id of Object.keys(env.data.transform)) result.push([query, id]);
	}
	return result;
}
//#endregion
//#region src/node/rpc/functions/get-module-transform-info.ts
const getModuleTransformInfo = defineRpcFunction({
	name: "vite-plugin-inspect:get-module-transform-info",
	type: "query",
	setup: (devtoolsCtx) => {
		const ctx = getInspectContext(devtoolsCtx);
		return {
			handler: async (query, id) => ctx.queryEnv(query).getModuleTransformInfo(id),
			dump: { inputs: getAllModuleIds(ctx).map(([q, id]) => [q, id]) }
		};
	}
});
//#endregion
//#region src/node/rpc/functions/get-modules-list.ts
const getModulesList = defineRpcFunction({
	name: "vite-plugin-inspect:get-modules-list",
	type: "query",
	setup: (devtoolsCtx) => {
		const ctx = getInspectContext(devtoolsCtx);
		return {
			handler: async (query) => ctx.queryEnv(query).getModulesList(),
			dump: { inputs: getAllQueryEnvs(ctx).map((q) => [q]) }
		};
	}
});
//#endregion
//#region src/node/rpc/functions/get-plugin-metrics.ts
const getPluginMetrics = defineRpcFunction({
	name: "vite-plugin-inspect:get-plugin-metrics",
	type: "query",
	setup: (devtoolsCtx) => {
		const ctx = getInspectContext(devtoolsCtx);
		return {
			handler: async (query) => ctx.queryEnv(query).getPluginMetrics(),
			dump: { inputs: getAllQueryEnvs(ctx).map((q) => [q]) }
		};
	}
});
//#endregion
//#region src/node/rpc/functions/get-server-metrics.ts
const getServerMetrics = defineRpcFunction({
	name: "vite-plugin-inspect:get-server-metrics",
	type: "query",
	setup: (devtoolsCtx) => {
		const ctx = getInspectContext(devtoolsCtx);
		return {
			handler: async (query) => ctx.getViteContext(query.vite).data.serverMetrics || {},
			dump: { inputs: getAllQueryEnvs(ctx).map((q) => [q]) }
		};
	}
});
//#endregion
//#region src/node/rpc/index.ts
const rpcFunctions = [
	clearModuleTransform,
	getMetadata,
	getModulesList,
	getPluginMetrics,
	getModuleTransformInfo,
	defineRpcFunction({
		name: "vite-plugin-inspect:resolve-id",
		type: "query",
		setup: (devtoolsCtx) => {
			const ctx = getInspectContext(devtoolsCtx);
			return {
				handler: async (query, id) => ctx.queryEnv(query).resolveId(id),
				dump: { inputs: getAllModuleIds(ctx).map(([q, id]) => [q, id]) }
			};
		}
	}),
	getServerMetrics
];
//#endregion
//#region src/node/index.ts
const NAME = "vite-plugin-inspect";
function PluginInspect(options = {}) {
	const { dev = true, build = false } = options;
	if (!dev && !build) return { name: NAME };
	const ctx = new InspectContext(options);
	let onBuildEnd;
	const timestampRE = /\bt=\d{13}&?\b/;
	const trailingSeparatorRE = /[?&]$/;
	function setupMiddlewarePerf(ctx, middlewares) {
		let firstMiddlewareIndex = -1;
		middlewares.forEach((middleware, index) => {
			const { handle: originalHandle } = middleware;
			if (typeof originalHandle !== "function" || !originalHandle.name) return middleware;
			middleware.handle = function(...middlewareArgs) {
				var _req$url, _ref;
				let req;
				if (middlewareArgs.length === 4) [, req] = middlewareArgs;
				else [req] = middlewareArgs;
				const start = Date.now();
				const url = (_req$url = req.url) === null || _req$url === void 0 ? void 0 : _req$url.replace(timestampRE, "").replace(trailingSeparatorRE, "");
				(_ref = ctx.data.serverMetrics.middleware)[url] ?? (_ref[url] = []);
				if (firstMiddlewareIndex < 0) firstMiddlewareIndex = index;
				if (index === firstMiddlewareIndex) ctx.data.serverMetrics.middleware[url] = [];
				const result = originalHandle.apply(this, middlewareArgs);
				Promise.resolve(result).then(() => {
					var _metrics$at;
					const total = Date.now() - start;
					const metrics = ctx.data.serverMetrics.middleware[url];
					ctx.data.serverMetrics.middleware[url].push({
						self: metrics.length ? Math.max(total - (((_metrics$at = metrics.at(-1)) === null || _metrics$at === void 0 ? void 0 : _metrics$at.total) || 0), 0) : total,
						total,
						name: originalHandle.name
					});
				});
				return result;
			};
			Object.defineProperty(middleware.handle, "name", {
				value: originalHandle.name,
				configurable: true,
				enumerable: true
			});
			return middleware;
		});
	}
	function configureServer(server) {
		Object.values(server.environments).forEach((env) => {
			const envCtx = ctx.getEnvContext(env);
			const _invalidateModule = env.moduleGraph.invalidateModule;
			env.moduleGraph.invalidateModule = function(...args) {
				const mod = args[0];
				if (mod === null || mod === void 0 ? void 0 : mod.id) envCtx.invalidate(mod.id);
				return _invalidateModule.apply(this, args);
			};
		});
		const base = (options.base ?? server.config.base) || "/";
		server.middlewares.use(`${base}__inspect`, (req, res) => {
			var _req$url2;
			const newUrl = ((_req$url2 = req.url) === null || _req$url2 === void 0 ? void 0 : _req$url2.replace(/^\/?/, `${base}.vite-inspect/`)) ?? `${base}.vite-inspect/`;
			res.writeHead(302, { Location: newUrl });
			res.end();
		});
		server.middlewares.use(`${base}.vite-inspect`, sirv(DIR_CLIENT, {
			single: true,
			dev: true
		}));
	}
	function setupDevTools(ctx, base) {
		return {
			capabilities: {
				dev: true,
				build: true
			},
			async setup(devtoolsCtx) {
				setInspectContext(devtoolsCtx, ctx);
				for (const fn of rpcFunctions) devtoolsCtx.rpc.register(fn);
				if (devtoolsCtx.docks) devtoolsCtx.docks.register({
					id: "vite-plugin-inspect",
					title: "Inspect",
					icon: "ph:magnifying-glass-duotone",
					type: "iframe",
					url: `${base}.vite-inspect/`
				});
				devtoolsCtx.views.hostStatic(`${base}.vite-inspect/`, DIR_CLIENT);
				if (devtoolsCtx.viteServer) {
					const debouncedBroadcast = debounce(() => {
						devtoolsCtx.rpc.broadcast({
							method: "vite-plugin-inspect:on-module-updated",
							args: []
						});
					}, 100);
					devtoolsCtx.viteServer.middlewares.use((req, res, next) => {
						debouncedBroadcast();
						next();
					});
				}
			}
		};
	}
	return {
		name: NAME,
		enforce: "pre",
		devtools: setupDevTools(ctx, options.base || "/"),
		apply(_, { command }) {
			if (command === "serve" && dev) return true;
			if (command === "build" && build) return true;
			return false;
		},
		configResolved(config) {
			config.plugins.forEach((plugin) => hijackPlugin(plugin, ctx));
			const _createResolver = config.createResolver;
			config.createResolver = function(...args) {
				const _resolver = _createResolver.apply(this, args);
				return async function(...args) {
					const id = args[0];
					const aliasOnly = args[2];
					const ssr = args[3];
					const start = Date.now();
					const result = await _resolver.apply(this, args);
					const end = Date.now();
					if (result && result !== id) {
						const pluginName = aliasOnly ? "alias" : "vite:resolve (+alias)";
						ctx.getViteContext(config).getEnvContext(ssr ? "ssr" : "client").recordResolveId(id, {
							name: pluginName,
							result,
							start,
							end
						});
					}
					return result;
				};
			};
			if (build) {
				const buildGenerator = createBuildGenerator(ctx);
				onBuildEnd = createEnvOrderHooks(Object.keys(config.environments), {
					async onFirst() {
						await buildGenerator.setupOutputDir();
					},
					async onEach(pluginCtx) {
						await buildGenerator.generateForEnv(pluginCtx);
					},
					async onLast(pluginCtx) {
						await buildGenerator.generateMetadata();
						const dir = buildGenerator.getOutputDir();
						pluginCtx.environment.logger.info(`${c.green("Inspect report generated at")}  ${c.dim(dir)}`);
					}
				});
			}
		},
		configureServer(server) {
			configureServer(server);
			return () => {
				setupMiddlewarePerf(ctx.getViteContext(server.config), server.middlewares.stack);
			};
		},
		load: {
			order: "pre",
			handler(id) {
				var _ctx$getEnvContext;
				(_ctx$getEnvContext = ctx.getEnvContext(this.environment)) === null || _ctx$getEnvContext === void 0 || _ctx$getEnvContext.invalidate(id);
				return null;
			}
		},
		hotUpdate({ modules }) {
			const ids = modules.map((module) => module.id);
			this.environment.hot.send({
				type: "custom",
				event: "vite-plugin-inspect:update",
				data: { ids }
			});
		},
		async buildEnd() {
			onBuildEnd === null || onBuildEnd === void 0 || onBuildEnd(this.environment.name, this);
		},
		sharedDuringBuild: true
	};
}
//#endregion
export { PluginInspect as t };
